Source/Artifacts/Artifacts.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Subjects;
+using Microsoft.Extensions.Logging;
+
+namespace Dolittle.SDK.Artifacts
+{
+    /// <summary>
+    /// Represents an implementation of <see cref="IArtifacts" />.
+    /// </summary>
+    public class Artifacts : IArtifacts, IDisposable
+    {
+        readonly ReplaySubject<ArtifactAssociation> _registered;
+        readonly BehaviorSubject<IDictionary<Type, Artifact>> _associations;
+        readonly ILogger<Artifacts> _logger;
+        bool _disposed;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Artifacts"/> class.
+        /// </summary>
+        /// <param name="associations">The <see cref="IDictionary{TKey, TValue}"> artifact associations </see>.</param>
+        /// <param name="logger">The <see cref="ILogger" />.</param>
+        public Artifacts(IDictionary<Type, Artifact> associations, ILogger<Artifacts> logger)
+        {
+            _registered = new ReplaySubject<ArtifactAssociation>();
+            _associations = new BehaviorSubject<IDictionary<Type, Artifact>>(new Dictionary<Type, Artifact>());
+            _logger = logger;
+
+            foreach ((var type, var artifact) in associations)
+            {
+                _registered.OnNext(new ArtifactAssociation(type, artifact));
+            }
+
+            _registered.Subscribe(AddAssociation);
+        }
+
+        /// <inheritdoc />
+        public void Associate(Type type, Artifact artifact)
+            => _registered.OnNext(new ArtifactAssociation(type, artifact));
+
+        /// <inheritdoc />
+        public void Associate(Type type, ArtifactId artifactId)
+            => Associate(type, new Artifact(artifactId, Generation.First));
+
+        /// <inheritdoc />
+        public void Associate(Type type, ArtifactId artifactId, Generation generation)
+            => Associate(type, new Artifact(artifactId, generation));
```

Comment:
Do we really need these methods? This seems like some conversions that should be done elsewhere. This isn't really a thing that developers will see, right?

Reply:
Kinda related to above comment. No we don't need the functionality here, but we should be able to register event types (artifacts) by just passing in a guid string.

Reply:
Yes, but I think that is just as well served with builder methods to handle these cases. And in any case, you don't register artifacts, but event types (the only one right now).

Reply:
True. The builder can be more flexible. But still does not answer the question of whether the Artifact should be abstract or not

Reply:
I do though agree that it should be abstract. Let me expand it with EventType and see where that goes :)